### PR TITLE
XWIKI-21034: Write functional tests and Page Object for Quick Actions

### PR DESCRIPTION
**Pull request by douakli**
Jira: https://jira.xwiki.org/browse/XWIKI-21034

This pull request adds a a page object and a first functional test for Quick Actions.

### CODE REVIEW

File: xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-pageobjects/src/main/java/org/xwiki/ckeditor/test/po/QuickActionDropdown.java [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+            }
+        }
+    }
+    
+    /**
+     * Submits the selected Quick Action.
+     */
+    public void submit()
+    {
+        if (this.getDropdown() != null) {
```
Comment (mflorea) [inline xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-pageobjects/src/main/java/org/xwiki/ckeditor/test/po/QuickActionDropdown.java:122]: This IF doesn't make sense because if the test calls ``submit`` then it means it expects the drop down to be displayed.

File: xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-pageobjects/src/main/java/org/xwiki/ckeditor/test/po/QuickActionDropdown.java [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+            return null;
+        }
+    }
+    
+    
+    /**
+     * Open the Quick Action drop-down.
+     * 
+     */
+    public void open()
```
Comment (mflorea) [inline xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-pageobjects/src/main/java/org/xwiki/ckeditor/test/po/QuickActionDropdown.java:79]: I'm wondering if it's not better to have a ``CKEditor.openQuickActions()`` method instead. This would allow you to do the wait in the ``QuickActionDropdown`` constructor and to also get a reference to the underlying ``WebElement`` instance there.

File: xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-pageobjects/src/main/java/org/xwiki/ckeditor/test/po/QuickActionDropdown.java
Comment (mflorea) [inline xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-pageobjects/src/main/java/org/xwiki/ckeditor/test/po/QuickActionDropdown.java:1]: The plural ``Actions`` sounds better to me, as this is a drop down that lists quick actions.

File: xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-pageobjects/src/main/java/org/xwiki/ckeditor/test/po/QuickActionDropdown.java [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+        }
+    }
+    
+    /**
+     * Finds the element enclosing given text.
+     * @param tag - search query
+     * @param content - the text to be found in the element
+     * @return the corresponding element
+     */
+    public boolean findElementsWithContent(String tag, String content)
```
Comment (mflorea) [inline xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-pageobjects/src/main/java/org/xwiki/ckeditor/test/po/QuickActionDropdown.java:134]: It seems this method looks in the edited content, so it doesn't seem related to the quick actions drop down page object. Also the method name suggests it return an element, but in fact it returns a boolean for which ``hasElementsWithContent`` is a better name. Then I'm not sure you need this method when you have ``RichTextAreaElement#getContent()``.

File: xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-docker/src/test/it/org/xwiki/ckeditor/test/ui/QuickActionIT.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+
+/**
+ * All functional tests for Quick Actions.
+ *
+ * @version $Id$
+ * @since 15.5RC1
+ */
+@UITest
+public class QuickActionIT
```
Comment (mflorea) [inline xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-docker/src/test/it/org/xwiki/ckeditor/test/ui/QuickActionIT.java:41]:
```suggestion
public class QuickActionsIT 
```

File: xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-docker/src/test/it/org/xwiki/ckeditor/test/ui/QuickActionIT.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+@UITest
+public class QuickActionIT 
+{
+    
+    /**
+     * Test string that will be inserted to check formatting.
+     */
+    private static final String TEST_TEXT = "Hello, world!";
+
+    @BeforeEach
```
Comment (mflorea) [inline xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-docker/src/test/it/org/xwiki/ckeditor/test/ui/QuickActionIT.java:49]: The browser is not closed after each test so the user remains logged in. Thus you need to login the user only once, at the start.

```suggestion
    @BeforeAll
```

File: xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-docker/src/test/it/org/xwiki/ckeditor/test/ui/QuickActionIT.java [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+        // Run the first quick action
+        QuickActionDropdown dropdown = new QuickActionDropdown(editor);
+        
+        dropdown.open();
+        dropdown.sendKeys("heading 1");
+        dropdown.submit();
+        
+        editor.getRichTextArea().sendKeys(TEST_TEXT);
+        assert dropdown.findElementsWithContent("h1", TEST_TEXT);
+        editPage.clickSaveAndView();
```
Comment (mflorea) [inline xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-docker/src/test/it/org/xwiki/ckeditor/test/ui/QuickActionIT.java:79]:
```suggestion
        WYSIWYGEditPage editPage = WYSIWYGEditPage.gotoPage(testReference);
        CKEditor editor = new CKEditor("content").waitToLoad();
        RichTextAreaElement textArea = editor.getRichTextArea();
        
        textArea.sendKeys("/hea");
        editor.getQuickActionsDropDown().waitForQuickActionSelected("Heading 1");
        textArea.sendKeys(Keys.ENTER, "TEST_TEXT");
        assertEquals("...", textArea.getContent());
```

**Review by mflorea — CHANGES_REQUESTED** [review]

File: xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-docker/src/test/it/org/xwiki/ckeditor/test/ui/QuickActionIT.java [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+import org.xwiki.test.docker.junit5.UITest;
+import org.xwiki.test.ui.TestUtils;
+import org.xwiki.test.ui.po.editor.WYSIWYGEditPage;
+
+
+/**
+ * All functional tests for Quick Actions.
+ *
+ * @version $Id$
+ * @since 15.5RC1
```
Comment (manuelleduc) [inline xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-docker/src/test/it/org/xwiki/ckeditor/test/ui/QuickActionIT.java:38]: Note: 15.5RC1 has been released, the `@since` will need to be updated to 15.5 or 15.6RC1 before merging.

Comment (douakli) [reply xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-docker/src/test/it/org/xwiki/ckeditor/test/ui/QuickActionIT.java:38]: Thank you!

File: xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-pageobjects/src/main/java/org/xwiki/ckeditor/test/po/QuickActionDropdown.java (hunk 2 above)
Comment (douakli) [reply xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-pageobjects/src/main/java/org/xwiki/ckeditor/test/po/QuickActionDropdown.java:79]: I moved it as you suggested, but I'm not sure it is necessary since as you mentioned, the user does not wait for the dropdown to show

File: xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-pageobjects/src/main/java/org/xwiki/ckeditor/test/po/QuickActionsDropdown.java [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+    }
+    
+    /**
+     * Waits for the given Quick Action to be selected.
+     * @param label - The name of the action
+     */
+    public void waitForQuickActionSelected(String label)
+    {
+        WebElement item = getDriver().findElement(By.className("cke_autocomplete_selected"));
+        item.findElement(By.xpath("//*[text() = '" + label + "']"));
```
Comment (mflorea) [inline xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-pageobjects/src/main/java/org/xwiki/ckeditor/test/po/QuickActionsDropdown.java:52]: Could be written in a single line.

```suggestion
        getDriver().findElement(By.xpath("//*[contains(@class, 'cke_autocomplete_selected')]//*[. = '" + label + "']"));
```

File: xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-docker/src/test/it/org/xwiki/ckeditor/test/ui/QuickActionsIT.java [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+        textArea.sendKeys("hea");
+        qa.waitForQuickActionSelected("Heading 1");
+        textArea.sendKeys(Keys.ENTER);
+        qa.waitForQuickActionSubmitted();
+        
+        // Write some text
+        textArea.sendKeys(TEST_TEXT);
+        
+        assert textArea.getContent().contains("<h1>" + TEST_TEXT + "<br></h1>");
+        editPage.clickSaveAndView();
```
Comment (mflorea) [inline xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-docker/src/test/it/org/xwiki/ckeditor/test/ui/QuickActionsIT.java:91]: Does it bring any value to save the page? Are you testing something with this? (maybe it helps with debugging?)

File: xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-docker/src/test/it/org/xwiki/ckeditor/test/ui/QuickActionsIT.java [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+    }
+    
+    @Test
+    @Order(2)
+    void heading2(TestUtils setup, TestReference testReference) throws Exception
```
Comment (mflorea) [inline xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-docker/src/test/it/org/xwiki/ckeditor/test/ui/QuickActionsIT.java:96]: I think it was enough to write a single heading test (the one that is more complex). But let's keep them.

File: xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-docker/src/test/it/org/xwiki/ckeditor/test/ui/QuickActionsIT.java [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+
+        // Switch to another style
+        QuickActionsDropdown qa = editor.openQuickActionsDropdown();
+        textArea.sendKeys("hea");
+        qa.waitForQuickActionSelected("Heading 1");
+        textArea.sendKeys(Keys.ENTER);
+        qa.waitForQuickActionSubmitted();
+        
+        // Write some text
+        textArea.sendKeys(TEST_TEXT);
```
Comment (mflorea) [inline xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-docker/src/test/it/org/xwiki/ckeditor/test/ui/QuickActionsIT.java:161]: You could assert here the HTML and drop the first heading test. It's fine to have a "story", i.e. to test multiple things in functional tests (compared to unit tests, which are less expensive).

**Review by mflorea — APPROVED** [review]
Looks very good. Let me know if you have something more to add (we should merge tomorrow).

File: xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-docker/src/test/it/org/xwiki/ckeditor/test/ui/QuickActionsIT.java (hunk 9 above)
Comment (douakli) [reply xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-docker/src/test/it/org/xwiki/ckeditor/test/ui/QuickActionsIT.java:91]: I'm saving only to avoid the confirmation dialog before leaving the page when starting the next test. There probably is a better way to handle that indeed.

File: xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-docker/src/test/it/org/xwiki/ckeditor/test/ui/QuickActionsIT.java (hunk 10 above)
Comment (douakli) [reply xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-docker/src/test/it/org/xwiki/ckeditor/test/ui/QuickActionsIT.java:96]: Oh okay, you did say to test every manually defined action so I went for it :)

File: xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-docker/src/test/it/org/xwiki/ckeditor/test/ui/QuickActionsIT.java (hunk 9 above)
Comment (mflorea) [reply xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-docker/src/test/it/org/xwiki/ckeditor/test/ui/QuickActionsIT.java:91]: Ah, I see, this deserves a comment. Now, instead of saving you can also cancel :) which is less expensive so normally faster. You can also handle the confirmation modal with Selenium but it's probably easier to just cancel the edit.

File: xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-docker/pom.xml [hunk 12]
```diff
@@ -58,6 +58,20 @@
...
+      <artifactId>xwiki-platform-mentions-ui</artifactId>
+      <version>${project.version}</version>
+      <type>xar</type>
+    </dependency>
+    <dependency>
+      <!-- We need this to test the code snippet Quick Action. -->
+      <groupId>org.xwiki.platform</groupId>
+      <artifactId>xwiki-platform-rendering-macros</artifactId>
+      <version>${project.version}</version>
+      <type>pom</type>
```
Comment (mflorea) [inline xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-docker/pom.xml:73]: Depending on the POM looks strange. Don't you need to depend on ``xwiki-platform-rendering-macro-code`` instead?

File: xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-pageobjects/src/main/java/org/xwiki/ckeditor/test/po/CKEditor.java [hunk 13]
```diff
@@ -155,13 +155,24 @@ private void internalClickImageButton()
...
+    
+    /**
+     * Clicks close on a CKEditor dialog.
+     */
+    public void cancelDialog() 
+    {
+        WebElement cancelButton = getDriver().findElement(By.className("cke_dialog_ui_button_cancel"));
+        getDriver().waitUntilElementIsEnabled(cancelButton);
+        cancelButton.click();
+        getDriver().waitUntilElementDisappears(By.className("cke_dialog_contents"));
```
Comment (mflorea) [inline xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-pageobjects/src/main/java/org/xwiki/ckeditor/test/po/CKEditor.java:176]: These two methods fit better in a ``CKEditorDialog`` page object.

File: xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-pageobjects/src/main/java/org/xwiki/ckeditor/test/po/MacroDialogEditModal.java [hunk 14]
```diff
@@ -37,7 +37,8 @@ public class MacroDialogEditModal extends BaseElement
      */
     public MacroDialogEditModal waitUntilReady()
     {
-        getDriver().waitUntilElementIsVisible(By.cssSelector(".gadget-editor-modal .macro-name"));
+        getDriver().waitUntilElementIsVisible(
+                By.xpath("//*[contains(@class, 'editor-modal')]//*[contains(@class, 'macro-name')]"));
```
Comment (mflorea) [inline xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-pageobjects/src/main/java/org/xwiki/ckeditor/test/po/MacroDialogEditModal.java:41]: Would be good to add a comment explaining why we use the suffix ``editor-modal`` instead of the full class name. And you can use a CSS selector, ``[attr*=value]``, see https://developer.mozilla.org/en-US/docs/Web/CSS/Attribute_selectors .

    [class*='-editor-modal'] .macro-name

File: xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-pageobjects/src/main/java/org/xwiki/ckeditor/test/po/MacroDialogEditModal.java [hunk 15]
```diff
@@ -46,6 +47,7 @@ public MacroDialogEditModal waitUntilReady()
      */
     public void clickSubmit()
     {
-        getDriver().findElement(By.cssSelector(".gadget-editor-modal .modal-footer .btn-primary")).click();
+        getDriver().findElement(
+                By.xpath("//*[contains(@class, 'editor-modal')]//*[contains(@class, 'modal-footer')]//*[contains(@class, 'btn-primary')]")).click();
```
Comment (mflorea) [inline xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-pageobjects/src/main/java/org/xwiki/ckeditor/test/po/MacroDialogEditModal.java:51]: Same, the CSS selector might be simpler. XPath is useful when you need to match the content of an HTML element, or its parent or a sibling.

File: xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-pageobjects/src/main/java/org/xwiki/ckeditor/test/po/RichTextAreaElement.java [hunk 16]
```diff
@@ -155,4 +156,18 @@ private WebElement getActiveElement()
         getDriver().switchTo().frame(this.iframe);
         return getDriver().switchTo().activeElement();
     }
+    
+    /**
+     * Waits until a macro appears in the document.
+     */
+    public void waitUntilMacroAppears()
```
Comment (mflorea) [inline xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-pageobjects/src/main/java/org/xwiki/ckeditor/test/po/RichTextAreaElement.java:163]: I find this too specific. I'd rather have a more generic "wait until the content is re-loaded".

File: xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-pageobjects/src/main/java/org/xwiki/ckeditor/test/po/RichTextAreaElement.java [hunk 17]
```diff
@@ -155,4 +156,18 @@ private WebElement getActiveElement()
...
+    {
+        try {
+            getDriver().waitUntilElementDisappears(this.iframe, By.tagName("iframe"));;
```
Comment (mflorea) [inline xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-pageobjects/src/main/java/org/xwiki/ckeditor/test/po/RichTextAreaElement.java:166]: I don't understand this line. You pass the iframe as the parent and you're looking for a child element with name "iframe"? Doesn't make sense, because the iframe element doesn't have child nodes. There's also a double ``;`` at the end.

File: xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-docker/src/test/it/org/xwiki/ckeditor/test/ui/QuickActionsIT.java [hunk 18]
```diff
@@ -39,7 +43,9 @@
  * @version $Id$
  * @since 15.5
  */
-@UITest
+@UITest(extraJARs = {
+    "org.xwiki.platform:xwiki-platform-search-solr-query",
```
Comment (mflorea) [inline xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-docker/src/test/it/org/xwiki/ckeditor/test/ui/QuickActionsIT.java:47]: This needs a comment.

File: xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-docker/src/test/it/org/xwiki/ckeditor/test/ui/QuickActionsIT.java [hunk 19]
```diff
@@ -70,25 +88,35 @@ private static void createAndLoginStandardUser(TestUtils setup)
...
+        
+        // Switch back to paragraph
+        textArea.sendKeys("/");
+        qa = new AutocompleteDropdown();
+        textArea.sendKeys("parag");
+        qa.waitForItemSelected("Paragraph");
+        textArea.sendKeys(Keys.ENTER);
+        qa.waitForItemSubmitted();
+        
+        assert textArea.getContent().contains("<p>" + TEST_TEXT + "<br></p>");
```
Comment (mflorea) [inline xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-docker/src/test/it/org/xwiki/ckeditor/test/ui/QuickActionsIT.java:119]: I think we prefer the JUnit assertions (``Assertions.assertTrue``) because the behaviour of ``assert`` can be disabled, thus making the tests pass.

**Review by mflorea — CHANGES_REQUESTED** [review]

File: xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-docker/pom.xml (hunk 12 above)
Comment (tmortagne) [reply xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-docker/pom.xml:73]: Yes. The point of depending on a pom is to get the dependencies of a module without the module itself, but it does not seem to be something needed here.

That being said, it seems strange to need the code macro for a test. If you added it because it's used by a XAR you are testing then it should be added as dependency of that XAR instead.

File: xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-docker/src/test/it/org/xwiki/ckeditor/test/ui/QuickActionsIT.java (hunk 19 above)
Comment (tmortagne) [reply xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-docker/src/test/it/org/xwiki/ckeditor/test/ui/QuickActionsIT.java:119]: In general `assert` is generally more something mean to be used at runtime, as a checkpoint (but that's not really something we do in XWiki).

File: xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-pageobjects/src/main/java/org/xwiki/ckeditor/test/po/RichTextAreaElement.java (hunk 16 above)
Comment (douakli) [reply xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-pageobjects/src/main/java/org/xwiki/ckeditor/test/po/RichTextAreaElement.java:163]: I'm not sure how to tell that the content has been reloaded. The content refreshes but I believe the ckeditor widget take some time to appear?

File: xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-pageobjects/src/main/java/org/xwiki/ckeditor/test/po/RichTextAreaElement.java (hunk 17 above)
Comment (douakli) [reply xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-pageobjects/src/main/java/org/xwiki/ckeditor/test/po/RichTextAreaElement.java:166]: Thank you, I'm not even sure how the test can pass despite that mistake

Comment (douakli) [reply xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-pageobjects/src/main/java/org/xwiki/ckeditor/test/po/RichTextAreaElement.java:166]: Correction, the test fails when I remove the parent, I'm confused

Comment (michitux) [reply xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-pageobjects/src/main/java/org/xwiki/ckeditor/test/po/RichTextAreaElement.java:166]: It passes because it waits for the iframe to not have a child with tag `iframe`, a condition that is instantly true.

**Review by mflorea — COMMENTED** [review]
Looks good, with some small comments.

**Review by mflorea — APPROVED** [review]
